code/__DEFINES/profile.dm

```diff
@@ -0,0 +1,29 @@
+#define PROFILE_START ;PROFILE_STORE = list();PROFILE_SET;
+#define PROFILE_STOP ;PROFILE_STORE = null;
+
+#define PROFILE_SET ;PROFILE_TIME = TICK_USAGE_REAL; PROFILE_LINE = __LINE__; PROFILE_FILE = __FILE__; PROFILE_SLEEPCHECK = world.time;
+
+#define PROFILE_TICK ;\
+	if (PROFILE_STORE) {\
+		var/PROFILE_TICK_USAGE_REAL = TICK_USAGE_REAL;\
+		if (PROFILE_SLEEPCHECK == world.time) {\
+			var/PROFILE_STRING = "[PROFILE_FILE]:[PROFILE_LINE] - [__FILE__]:[__LINE__]";\
+			var/list/PROFILE_ITEM = PROFILE_STORE[PROFILE_STRING];\
+			if (!PROFILE_ITEM) {\
+				PROFILE_ITEM = new(PROFILE_ITEM_LEN);\
+				PROFILE_STORE[PROFILE_STRING] = PROFILE_ITEM;\
+				PROFILE_ITEM[PROFILE_ITEM_TIME] = 0;\
+				PROFILE_ITEM[PROFILE_ITEM_COUNT] = 0;\
```

Comment:
what the absolute fuck

Reply:
It looks pretty self explanatory.

Did you have a question?

Reply:
Magic byond voodoo, how does it work?

Reply:
what, what are you wtfing over, the define format, or the contents of the define.

Reply:
Everything about this

Reply:
Ok, so it works like this:

Store the tick_usage, world.time, and current line and file (byond macros `__LINE__` and `__FILE__`). This works because defines inside defines are resolved at usage time, so `__LINE__` becomes the line the macro was used in. (check `PROFILE_SET`)

Then, in the profile_tick method, if world.time is still the same (meaning its still the same byond tick) use the stored tick_usage and the current tick_usage to figure out how many milliseconds have passed.

Make a list. with index 1 being the time, and index 2 being the amount of times this line of code was reached (if it doesn't already exist). Store this in the `PROFILE_STORE` list keyed to the old file and line stored by `PROFILE_SET`, as well as the current file and line.

Increment index 1 (`PROFILE_ENTRY_TIME`) by the millisecond value from above

Increment index 2 (`PROFILE_ENTRY_COST`) by a special formula that takes into account byond float issues. (Basically after 16 million we can't do += 1, because under floats that would round back to the original number, and instead we have to increment it by 2 every 2 calls (by abusing prob()), and after 32 million, we increment it by 4 every 4 calls, and on and on)

Then call the `PROFILE_SET` macro again.

I also abuse `\` to format this define with newlines and shit. `\` at the end of a line merges the line with the next, as well as `;` and `{` `}` to let me do ifs and shit.

output looks like this: https://gist.github.com/MrStonedOne/06a275be398499db41da702dde7fc518 

you have to add `PROFILE_TICK` to every line or between every line, and it will automatically store the amount of time between profile_tick usages in a keyed list "[from_file]:[from_line] - [to_file]:[to_line]" basically. example: https://github.com/tgstation/tgstation/commit/ef876c6fdb153e97d6ff954722dfde9c78eff315#diff-7664a9b9849fbd5b5d688654469b26d9R434

Reply:
>Magic byond voodoo

Moron.

Reply:
remie knows real voodoo`......`

Reply:
That is Pretty Cool, I like it.

Reply:
Clever

Reply:
One too many dots @duncathan

Reply:
only sometimes